salt/utils/json.py

```diff
@@ -27,10 +27,48 @@ def __split(raw):
 def find_json(raw):
     """
     Pass in a raw string and load the json when it starts. This allows for a
-    string to start with garbage and end with json but be cleanly loaded
+    string to start or end with garbage but the JSON be cleanly loaded
     """
     ret = {}
     lines = __split(raw)
+    lengths = list(map(len, lines))
+    starts = []
+    ends = []
+
+    # Search for possible starts and ends of the json fragments
+    for ind, _ in enumerate(lines):
+        line = lines[ind].lstrip()
+        line = line[0] if line else line
+        if line == "{" or line == "[":
+            starts.append((ind, line))
+        if line == "}" or line == "]":
+            ends.append((ind, line))
+
+    # List all the possible pairs of starts and ends,
+    # and fill the length of each block to sort by size after
+    starts_ends = []
+    for start, start_br in starts:
+        for end, end_br in reversed(ends):
+            if end > start and (
+                (start_br == "{" and end_br == "}")
+                or (start_br == "[" and end_br == "]")
+            ):
+                starts_ends.append((start, end, sum(lengths[start : end + 1])))
+
+    # Iterate through all the possible pairs starting from the largest
+    starts_ends.sort(key=lambda x: (x[2], x[1] - x[0], x[0]), reverse=True)
```

Comment:
I don't think the largest json is necessarily the correct one. I think that the first one is probably what should returned. Either way it should probably be documented in the docstring.

Reply:
No, the largest JSON is the correct one. A JSON can contain an unspecified number of valid JSON objects, we cannot return anything but the largest JSON. This is exactly the original behavior, so there's no need to change the docstring.